### ISSUE TITLE
relax some installation versions in setup.py

### ISSUE THREAD

**Issue by rizac** [gem/oq-engine#7260]
Hi all

(sorry if the problem is already mentioned or fixed, I did not find any in the issues list)

TL/DR:

I would like to ask if it was possible to relax some installation conditions in `setup.py`. I understand that specific version ranges are useful for end users, but they make it hard for developers who use OpenQuake as dependency for other programs on top of it. 

I think that there might be plenty of alternative solutions that do not interfere with the stability of the program or the CI workflow and could be beneficial for both end users **and** developers, e.g.,  specify a minimum version only (typical approach) and/or play around with `install_requires` and  `extras_requires` in setup.py to allow specific installations, e.g. something like `pip install .[without-django]` for developers, and/or relaxing some conditions for those packages not strictly required for computation (matplotlib and Django, I presume), and so on.

---

Long explanation (for those who want to read):

Years ago we implemented a [web service based on the strong motion toolkit and openquake](https://github.com/rizac/eGSIM). It uses Django specifically because it was already a dependency of Openquake. In the long run (e.g., maintenance) this turned out to be a hindrance rather than an advantage. We sometimes face web-related security flaws that need upgrades to new versions of Django, which are urgent and most-likely completely unrelated with the core functionality of OpenQuake. Yet, we must stick to the version that OpenQuake uses, because when we update our requirements.txt, reinstalling later we get errors of this type:
```
ERROR: Cannot install Django>2.3 and openquake-engine==3.11.3 because these package versions have conflicting dependencies.

The conflict is caused by:
    The user requested Django>2.3
    openquake-engine 3.11.3 depends on django<2.3 and >=1.11

To fix this you could try to:
1. loosen the range of package versions you've specified
2. remove package versions to allow pip attempt to solve the dependency conflict

ERROR: ResolutionImpossible: for help visit https://pip.pypa.io/en/latest/user_guide/#fixing-conflicting-dependencies
```
In general, I find enlightening this comment (details [here](https://github.com/ckan/ckan/issues/4590#issuecomment-448204838)):

    install_requires in setup.py is convenient for end users, it makes sure to grab the 
    right versions of dependencies when using pip or setup.py install.
    At the same time install_requires is very painful for developers because it prevents them 
    from even running our code if any package is installed with any version conflict, 
    even if we're just testing that a newer dependency works properly.

**Comment by micheles** [comment]
I see your point but I do not see what we can do on our side; we *need* specific versions for testing purposes. My suggestion is that when you install the engine you do a pre-install step (for instance a sed on the setup.py) to relax the dependencies and that's it. 

BTW, about Django, we recently upgraded to version 3.2.

**Comment by rizac** [comment]
Thank you for the reply.

As said, I did not mean to force you to remove the version you need for your tests. 

I was thinking about something like playing around with `install_requires` and `extras_require` in `setup.py`, allowing **both** the normal case and extra flexibility  for developers. E.g. I know it might not be possible, but what about allow users to do a `pip install -e .` or `pip install -e ".[gui]"` after e.g., removing Django from the install_requires and specifying in `setup.py`:

```
    extras_require={
        'gui': [
            'django>2.6.1,<=3.5'
        ],
    },
```
(of course, assuming django is required only for GUI purposes in Openquake and not for the seismic core computations. Same e.g. for matplotlib, often a pain in the neck but not required by some other packages I used in the past).


BTW: I know that you upgraded to Django, thank you. But this does not solve the problem of next potential issue: a web related flaw might happen in two days and the next Openquake release with required Django versions in two months. It is unmaintainable this way. Also consider that we must stick to a previous version of OpenQuake and not the latest one, because the recent refactoring of OpenQuake can not be implemented in our web app for lack of human resources (and this is our fault obviously, but just FYI because it is unfortunately a quite common scenario)
